File: plugin/src/main/java/com/stratio/cassandra/lucene/column/ColumnsMapper.java
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene.column;

import com.stratio.cassandra.lucene.IndexException;
import org.apache.cassandra.config.ColumnDefinition;
import org.apache.cassandra.db.marshal.*;
import org.apache.cassandra.db.rows.Cell;
import org.apache.cassandra.db.rows.ComplexColumnData;
import org.apache.cassandra.db.rows.Row;
import org.apache.cassandra.serializers.CollectionSerializer;
import org.apache.cassandra.serializers.ListSerializer;
import org.apache.cassandra.serializers.MapSerializer;
import org.apache.cassandra.utils.ByteBufferUtil;

import java.nio.ByteBuffer;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;
import com.stratio.cassandra.lucene.core.column.*;

import static org.apache.cassandra.transport.Server.CURRENT_VERSION;

/**
 * Class for several regular column mappings between Cassandra and Lucene.
 *
 * @author Andres de la Pena {@literal <adelapena@stratio.com>}
 */
public class ColumnsMapper {

    /**
     * Returns the {@link Column}s contained in the specified {@link Row}.
     *
     * @param row a {@link Row}
     * @return the columns
     */
    public Columns columns(Row row) {
        Columns columns = Columns.build();
        for (ColumnDefinition columnDefinition : row.columns()) {
            if (columnDefinition.isComplex()) {
                columns = columns.add(columns(row.getComplexColumnData(columnDefinition)));
            } else {
                columns = columns.add(columns(row.getCell(columnDefinition)));
            }
        }
        return columns;
    }

    private Columns columns(ComplexColumnData complexColumnData) {
        Columns columns = Columns.build();
        if (complexColumnData != null) {
            for (Cell cell : complexColumnData) {
                columns = columns.add(columns(cell));
            }
        }
        return columns;
    }

    private Columns columns(Cell cell) {
        Columns columns = Columns.build();
        if (cell != null) {
            boolean isTombstone = cell.isTombstone();
            ColumnDefinition columnDefinition = cell.column();
            String name = columnDefinition.name.toString();
            AbstractType<?> type = cell.column().type;
            ByteBuffer value = cell.value();
            ColumnAdder adder = columns.adder(name, cell.localDeletionTime());
            if (type.isCollection() && !type.isFrozenCollection()) {
                CollectionType<?> collectionType = (CollectionType<?>) type;
                switch (collectionType.kind) {
                    case SET: {
                        type = collectionType.nameComparator();
                        value = cell.path().get(0);
                        columns = columns.add(columns(isTombstone, adder, type, value));
                        break;
                    }
                    case LIST: {
                        type = collectionType.valueComparator();
                        columns = columns.add(columns(isTombstone, adder, type, value));
                        break;
                    }
                    case MAP: {
                        type = collectionType.valueComparator();
                        ByteBuffer keyValue = cell.path().get(0);
                        AbstractType<?> keyType = collectionType.nameComparator();
                        String nameSuffix = keyType.compose(keyValue).toString();
                        columns = columns.add(columns(isTombstone, adder.withMapName(nameSuffix), type, value));
                        break;
                    }
                    default: {
                        throw new IndexException("Unknown collection type {}", collectionType.kind);
                    }
                }
            } else {
                columns = columns.add(columns(isTombstone, adder, type, value));
            }
        }
        return columns;
    }

    private Columns columns(boolean isTombstone,
                            ColumnAdder adder,
                            AbstractType type,
                            ByteBuffer value) {
        Columns columns = Columns.build();
        if (type.isCollection()) {
            value = ByteBufferUtil.clone(value);
            CollectionType<?> collectionType = (CollectionType<?>) type;
            switch (collectionType.kind) {
                case SET: {
                    AbstractType<?> nameType = collectionType.nameComparator();
                    if (isTombstone) {
                        columns = adder.add();
                    } else {
                        int colSize = CollectionSerializer.readCollectionSize(value, CURRENT_VERSION);
                        for (int j = 0; j < colSize; j++) {
                            ByteBuffer itemValue = CollectionSerializer.readValue(value, CURRENT_VERSION);
                            columns = columns.add(columns(false, adder, nameType, itemValue));
                        }
                    }
                    break;
                }
                case LIST: {
                    AbstractType<?> valueType = collectionType.valueComparator();
                    if (isTombstone) {
                        columns = adder.add();
                    } else {
                        int colSize = ListSerializer.readCollectionSize(value, CURRENT_VERSION);
                        for (int j = 0; j < colSize; j++) {
                            ByteBuffer itemValue = CollectionSerializer.readValue(value, CURRENT_VERSION);
                            columns = columns.add(columns(false, adder, valueType, itemValue));
                        }
                    }
                    break;
                }
                case MAP: {
                    AbstractType<?> keyType = collectionType.nameComparator();
                    AbstractType<?> valueType = collectionType.valueComparator();
                    if (isTombstone) {
                        columns = adder.add();
                    } else {
                        int colSize = MapSerializer.readCollectionSize(value, CURRENT_VERSION);
                        for (int j = 0; j < colSize; j++) {
                            ByteBuffer mapKey = MapSerializer.readValue(value, CURRENT_VERSION);
                            ByteBuffer mapValue = MapSerializer.readValue(value, CURRENT_VERSION);
                            String itemName = keyType.compose(mapKey).toString();
                            collectionType.nameComparator();
                            columns = columns.add(columns(false, adder.withMapName(itemName), valueType, mapValue));
                        }
                    }
                    break;
                }
                default: {
                    throw new IndexException("Unknown collection type {}", collectionType.kind);
                }
            }
        } else if (type instanceof UserType) {
            UserType userType = (UserType) type;
            ByteBuffer[] values = userType.split(value);
            for (int i = 0; i < userType.fieldNames().size(); i++) {
                String itemName = userType.fieldNameAsString(i);
                AbstractType<?> itemType = userType.fieldType(i);
                if (isTombstone) {
                    columns = adder.withUDTName(itemName).add();
                } else if (values[i] != null) { // This only occurs in UDT not fully composed
                    columns = columns.add(columns(false, adder.withUDTName(itemName), itemType, values[i]));
                }
            }
        } else if (type instanceof TupleType) {
            TupleType tupleType = (TupleType) type;
            ByteBuffer[] values = tupleType.split(value);
            for (Integer i = 0; i < tupleType.size(); i++) {
                String itemName = i.toString();
                AbstractType<?> itemType = tupleType.type(i);
                if (isTombstone) {
                    columns = adder.withUDTName(itemName).add();
                } else {
                    columns = columns.add(columns(false, adder.withUDTName(itemName), itemType, values[i]));
                }
            }
        } else {
            if (value != null) {
                columns = adder.add(value(value, type));
            }
        }
        return columns;
    }

    public static <T> Column<?> column(String name, ByteBuffer bb, AbstractType<T> type) {
        return Column.build(name, value(bb, type));
    }

    public static Object value(ByteBuffer bb, AbstractType<?> type) {
        if (type instanceof SimpleDateType) {
            Locale locale = Locale.getDefault(Locale.Category.FORMAT);
            Calendar calendar = Calendar.getInstance(TimeZone.getDefault(), locale);
            long timestamp = SimpleDateType.instance.toTimeInMillis(bb);
            timestamp -= calendar.getTimeZone().getOffset(timestamp);
            return new Date(timestamp);
        }
        return type.compose(bb);
    }

}

File: plugin/src/main/scala/com/stratio/cassandra/lucene/column/ColumnsMapper.scala
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene.column

import java.math.{BigDecimal, BigInteger}
import java.net.InetAddress
import java.nio.ByteBuffer
import java.util.{Date, UUID}

import com.stratio.cassandra.lucene.IndexException
import org.apache.cassandra.config.CFMetaData
import org.apache.cassandra.db.marshal._
import org.apache.cassandra.db.rows.{Cell, ComplexColumnData, Row}
import org.apache.cassandra.serializers.CollectionSerializer
import org.apache.cassandra.transport.Server._
import org.apache.cassandra.utils.ByteBufferUtil

import scala.annotation.tailrec
import scala.collection.JavaConversions._

/** Maps Cassandra rows to [[Columns]].
  *
  * @author Andres de la Pena `adelapena@stratio.com`
  */
object ColumnsMapper {

  /** Returns a [[Columns]] representing the specified row.
    *
    * @param row the Cassandra row to be mapped
    */
  def columns(row: Row): Columns = {
    row.columns().foldLeft(Columns())(
      (cs, columnDefinition) =>
        if (columnDefinition.isComplex)
          cs + columns(row.getComplexColumnData(columnDefinition))
        else
          cs + columns(row.getCell(columnDefinition))
    )
  }

  private[column] def columns(complexColumnData: ComplexColumnData): Columns = {
    complexColumnData.foldLeft(Columns())((cs, cell) => cs + columns(cell))
  }

  private[column] def columns(cell: Cell): Columns = {
    if (cell == null) return Columns()
    val isTombstone = cell.isTombstone
    val name = cell.column.name.toString
    val comparator = cell.column.`type`
    val value = cell.value
    val column = new Column(cellName = name, deletionTime = cell.localDeletionTime)
    comparator match {
      case setType: SetType[_] if !setType.isFrozenCollection =>
        val itemComparator = setType.nameComparator
        val itemValue = cell.path.get(0)
        columns(isTombstone, column, itemComparator, itemValue)
      case listType: ListType[_] if !listType.isFrozenCollection =>
        val itemComparator = listType.valueComparator
        columns(isTombstone, column, itemComparator, value)
      case mapType: MapType[_, _] if !mapType.isFrozenCollection =>
        val itemComparator = mapType.valueComparator
        val keyValue = cell.path.get(0)
        val keyComparator = mapType.nameComparator
        val nameSuffix = keyComparator.compose(keyValue).toString
        columns(isTombstone, column.withMapName(nameSuffix), itemComparator, value)
      case _ =>
        columns(isTombstone, column, comparator, value)
    }
  }

  private[column] def columns(
      isTombstone: Boolean,
      column: Column[_],
      abstractType: AbstractType[_],
      value: ByteBuffer): Columns = abstractType match {
    case setType: SetType[_] =>
      columns(isTombstone, column, setType, value)
    case listType: ListType[_] =>
      columns(isTombstone, column, listType, value)
    case mapType: MapType[_, _] =>
      columns(isTombstone, column, mapType, value)
    case userType: UserType =>
      columns(isTombstone, column, userType, value)
    case tupleType: TupleType =>
      columns(isTombstone, column, tupleType, value)
    case _ =>
      Columns(column.withValue(compose(value, abstractType)))
  }

  private[this] def columns(
      isTombstone: Boolean,
      column: Column[_],
      set: SetType[_],
      value: ByteBuffer): Columns = {
    if (isTombstone) return Columns(column)
    val nameType = set.nameComparator
    val bb = ByteBufferUtil.clone(value) // CollectionSerializer read functions are impure
    (0 until frozenCollectionSize(bb)).foldLeft(Columns())(
      (cs, n) => {
        val itemValue = frozenCollectionValue(bb)
        cs + columns(isTombstone, column, nameType, itemValue)
      })
  }

  private[this] def columns(
      isTombstone: Boolean,
      column: Column[_],
      list: ListType[_],
      value: ByteBuffer): Columns = {
    if (isTombstone) return Columns(column)
    val valueType = list.valueComparator
    val bb = ByteBufferUtil.clone(value) // CollectionSerializer read functions are impure
    (0 until frozenCollectionSize(bb)).foldLeft(Columns())(
      (cs, n) => {
        val itemValue = frozenCollectionValue(bb)
        cs + columns(isTombstone, column, valueType, itemValue)
      })
  }

  private[this] def columns(
      isTombstone: Boolean,
      column: Column[_],
      map: MapType[_, _],
      value: ByteBuffer): Columns = {
    if (isTombstone) return Columns(column)
    val itemKeysType = map.nameComparator
    val itemValuesType = map.valueComparator
    val bb = ByteBufferUtil.clone(value) // CollectionSerializer read functions are impure
    (0 until frozenCollectionSize(bb)).foldLeft(Columns())(
      (cs, n) => {
        val itemKey = frozenCollectionValue(bb)
        val itemValue = frozenCollectionValue(bb)
        val itemName = itemKeysType.compose(itemKey).toString
        cs + columns(isTombstone, column.withMapName(itemName), itemValuesType, itemValue)
      })
  }

  private[this] def columns(
      isTombstone: Boolean,
      column: Column[_],
      udt: UserType,
      value: ByteBuffer): Columns = {
    if (isTombstone) return Columns(column)
    val itemValues = udt.split(value)
    (0 until udt.fieldNames.size).foldLeft(Columns())(
      (cs, i) => {
        val itemName = udt.fieldNameAsString(i)
        val itemType = udt.fieldType(i)
        val itemValue = itemValues(i)
        if (isTombstone || itemValue == null) {
          cs + column.withUDTName(itemName)
        } else {
          cs + columns(isTombstone, column.withUDTName(itemName), itemType, itemValue)
        }
      })
  }

  private[this] def columns(
      isTombstone: Boolean,
      column: Column[_],
      tuple: TupleType,
      value: ByteBuffer): Columns = {
    if (isTombstone) return Columns(column)
    val itemValues = tuple.split(value)
    (0 until tuple.size).foldLeft(Columns())(
      (cs, i) => {
        val itemName = i.toString
        val itemType = tuple.`type`(i)
        val itemValue = itemValues(i)
        if (isTombstone || itemValue == null) {
          cs + column.withUDTName(itemName)
        } else {
          cs + columns(isTombstone, column.withUDTName(itemName), itemType, itemValue)
        }
      })
  }

  private[this] def frozenCollectionSize(bb: ByteBuffer): Int =
    CollectionSerializer.readCollectionSize(bb, CURRENT_VERSION)

  private[this] def frozenCollectionValue(bb: ByteBuffer): ByteBuffer =
    CollectionSerializer.readValue(bb, CURRENT_VERSION)

  def compose(bb: ByteBuffer, t: AbstractType[_]): Any = t match {
    case sdt: SimpleDateType => new Date(sdt.toTimeInMillis(bb))
    case _ => t.compose(bb)
  }

  ///////////////////////////////////////////////////////////////////////////
  // Validation
  ///////////////////////////////////////////////////////////////////////////

  def validate(
      metadata: CFMetaData,
      column: String,
      field: String,
      supportedTypes: java.util.List[Class[_]]) {

    val cellName = Column.parse(column).cellName
    val cellDefinition = metadata.getColumnDefinition(UTF8Type.instance.decompose(cellName))

    if (cellDefinition == null) {
      throw new IndexException(s"No column definition '$cellName' for mapper '$field'")
    }
    if (cellDefinition.isStatic) {
      throw new IndexException(s"Lucene indexes are not allowed on static columns as '$column'")
    }

    def checkSupported(t: AbstractType[_], mapper: String) {
      if (!supports(t, supportedTypes)) throw new IndexException(
        s"Type '$t' in column '$mapper' is not supported by mapper '$field'")
    }

    val cellType = cellDefinition.`type`
    val udtNames = Column.parse(column).udtNames
    if (udtNames.isEmpty) {
      checkSupported(cellType, cellName)
    } else {
      var col = Column.apply(cellName)
      var currentType = cellType
      for (i <- udtNames.indices) {
        col = col.withUDTName(udtNames(i))
        val mapperName = col.mapperName
        ColumnsMapper.childType(currentType, udtNames(i)) match {
          case None =>
            throw new IndexException(s"No column definition '$mapperName' for mapper '$field'")
          case Some(n) if i == udtNames.indices.last => checkSupported(n, mapperName)
          case Some(n) => currentType = n
        }
      }
    }
  }

  @tailrec
  def childType(parent: AbstractType[_], child: String): Option[AbstractType[_]] = parent match {
    case t: ReversedType[_] => childType(t.baseType, child)
    case t: SetType[_] => childType(t.nameComparator, child)
    case t: ListType[_] => childType(t.valueComparator, child)
    case t: MapType[_, _] => childType(t.valueComparator, child)
    case t: UserType =>
      (0 until t.fieldNames.size).find(t.fieldNameAsString(_) == child).map(t.fieldType)
    case t: TupleType => (0 until t.size).find(_.toString == child).map(t.`type`)
    case _ => None
  }

  @tailrec
  def supports(
      candidateType: AbstractType[_],
      supportedTypes: Seq[Class[_]]): Boolean = candidateType match {
    case t: ReversedType[_] => supports(t.baseType, supportedTypes)
    case t: SetType[_] => supports(t.getElementsType, supportedTypes)
    case t: ListType[_] => supports(t.getElementsType, supportedTypes)
    case t: MapType[_, _] => supports(t.getValuesType, supportedTypes)
    case _ =>
      val native = nativeType(candidateType)
      supportedTypes.exists(_ isAssignableFrom native)
  }

  def nativeType(validator: AbstractType[_]): Class[_] = validator match {
    case _: UTF8Type | _: AsciiType => classOf[String]
    case _: SimpleDateType | _: TimestampType => classOf[Date]
    case _: UUIDType | _: LexicalUUIDType | _: TimeUUIDType => classOf[UUID]
    case _: ShortType => classOf[java.lang.Short]
    case _: ByteType => classOf[java.lang.Byte]
    case _: Int32Type => classOf[Integer]
    case _: LongType => classOf[java.lang.Long]
    case _: IntegerType => classOf[BigInteger]
    case _: FloatType => classOf[java.lang.Float]
    case _: DoubleType => classOf[java.lang.Double]
    case _: DecimalType => classOf[BigDecimal]
    case _: BooleanType => classOf[java.lang.Boolean]
    case _: BytesType => classOf[ByteBuffer]
    case _: InetAddressType => classOf[InetAddress]
    case _ => throw new IndexException(s"Unsupported Cassandra data type: ${validator.getClass}")
  }

}
